In [1]:
import os
from dotenv import load_dotenv

load_dotenv()

os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")
os.environ["ANTHROPIC_API_KEY"] = os.getenv("ANTHROPIC_API_KEY")
os.environ["GROQ_API_KEY"] = os.getenv("GROQ_API_KEY")

os.environ["LANGCHAIN_API_KEY"] = os.getenv("LANGCHAIN_API_KEY")
os.environ["LANGCHAIN_PROJECT"] = os.getenv("LANGCHAIN_PROJECT")
os.environ["LANGCHAIN_TRACING_V2"] = "true"

In [5]:
import asyncio
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_core.tools import FunctionTool

In [3]:
model_client = OpenAIChatCompletionClient(
    model='gpt-4o-mini',
    temperature=0.3
)

In [20]:
def reverse_string(text: str) -> str:
    '''
    Reverse the given text

    input:str

    output:str

    The reverse string is returned.
    '''
    return text[::-1]

In [21]:
reverse_tool = FunctionTool(reverse_string,description='A tool to reverse a string')
reverse_tool

In [22]:
agent = AssistantAgent(
    name="ReverseStringAgent",
    model_client= model_client,
    system_message='You are a helpful assistant that can reverse string using reverse_string tool. Give the result with summary',
    tools=[reverse_tool],
    reflect_on_tool_use=True
)

In [23]:
result = await agent.run(task = 'Reverse the string "Hello, World!"')

In [24]:
result

TaskResult(messages=[TextMessage(id='71379955-db33-4d01-8474-c4c673e43e82', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 48, 58, 648510, tzinfo=datetime.timezone.utc), content='Reverse the string "Hello, World!"', type='TextMessage'), ToolCallRequestEvent(id='ce9f6d3a-bf79-4333-9625-dfd133a17351', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=79, completion_tokens=17), metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 48, 59, 575128, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_LHwcWAxOiaJopRkYwX7LRLSY', arguments='{"text":"Hello, World!"}', name='reverse_string')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='4e377412-bbf0-4c10-891b-f6d42db62902', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 48, 59, 575128, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string',

TaskResult(
    messages=[
        TextMessage(id='71379955-db33-4d01-8474-c4c673e43e82', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 48, 58, 648510, tzinfo=datetime.timezone.utc), content='Reverse the string "Hello, World!"', type='TextMessage'), ToolCallRequestEvent(id='ce9f6d3a-bf79-4333-9625-dfd133a17351', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=79, completion_tokens=17), metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 48, 59, 575128, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_LHwcWAxOiaJopRkYwX7LRLSY', arguments='{"text":"Hello, World!"}', name='reverse_string')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='4e377412-bbf0-4c10-891b-f6d42db62902', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 48, 59, 575128, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string', call_id='call_LHwcWAxOiaJopRkYwX7LRLSY', is_error=False)], type='ToolCallExecutionEvent'), TextMessage(id='b69402f7-6d99-44e3-a7b0-24821b28d5ad', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=71, completion_tokens=30), metadata={}, created_at=datetime.datetime(2025, 7, 20, 16, 49, 1, 73271, tzinfo=datetime.timezone.utc), content='The reversed string of "Hello, World!" is "!dlroW ,olleH". \n\n**Summary:** The original string has been successfully reversed.', type='TextMessage')], stop_reason=None)

In [25]:
print(result.messages[-1].content)

The reversed string of "Hello, World!" is "!dlroW ,olleH". 

**Summary:** The original string has been successfully reversed.
